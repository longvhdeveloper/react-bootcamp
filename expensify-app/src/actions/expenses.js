import uuid from 'uuid'

// add expense
export const addExpense = ({
  description = '',
  notes = '',
  amount = 0,
  createAt = 0,
} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    notes,
    amount,
    createAt,
  },
})

// remove expense
export const removeExpense = ({ id = '' } = {}) => {
  return {
    type: 'REMOVE_EXPENESE',
    id,
  }
}

// edit expense
export const editExpense = (id, updates = {}) => {
  return {
    type: 'EDIT_EXPENSE',
    id,
    updates,
  }
}
