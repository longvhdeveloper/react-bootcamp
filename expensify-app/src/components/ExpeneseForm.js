import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'

// const date = new Date()
const now = moment()
console.log(now.format('MMM Do, YYYY'))

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      description: props.expense ? props.expense.description : '',
      notes: props.expense ? props.expense.notes : '',
      amount: props.expense ? props.expense.amount / 100 : '',
      createAt: props.expense ? moment(props.expense.createAt) : moment(),
      calendarFocused: false,
      error: '',
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onNoteChange = (e) => {
    const notes = e.target.value
    this.setState(() => ({ notes }))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChange = (createAt) => {
    if (createAt) {
      this.setState(() => ({ createAt }))
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onSubmit = (e) => {
    e.preventDefault()

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount.' }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createAt: this.state.createAt.valueOf(),
        notes: this.state.notes,
      })
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p className="error">{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            value={this.state.description}
            type="text"
            placeholder="Description"
            onChange={this.onDescriptionChange}
          />
          <input
            value={this.state.amount}
            type="text"
            placeholder="Amount"
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            onChange={this.onNoteChange}
            placeholder="Notes for expense"
          >
            {this.state.notes}
          </textarea>
          <button>Save</button>
        </form>
      </div>
    )
  }
}
