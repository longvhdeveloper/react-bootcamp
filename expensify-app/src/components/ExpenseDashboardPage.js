import React from 'react'

const ExpenseDashboardPage = () => {
  return <div>This is from my dashboard page</div>
}

export default ExpenseDashboardPage
