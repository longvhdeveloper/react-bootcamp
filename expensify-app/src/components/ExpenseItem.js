import React from 'react'
import { Link } from 'react-router-dom'

export const ExpneseItem = ({ expense }) => {
  return (
    <tr>
      <td>{expense.description}</td>
      <td>{expense.notes}</td>
      <td>{expense.amount}</td>
      <td>{expense.createAt}</td>
      <td>
        <Link to={`/edit/${expense.id}`}>Edit</Link>
      </td>
    </tr>
  )
}

export default ExpneseItem
