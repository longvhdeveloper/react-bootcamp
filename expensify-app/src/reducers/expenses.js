// expense reducer
const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense]
    case 'REMOVE_EXPENESE':
      return state.filter((expense) => expense.id != action.id)
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates,
          }
        } else {
          return expense
        }
      })
    default:
      return state
  }
}
