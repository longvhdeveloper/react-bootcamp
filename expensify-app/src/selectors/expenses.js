import moment from 'moment'

// get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createAtMoment = moment(expense.createAt)
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createAtMoment, 'date')
        : true

      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createAtMoment, 'date')
        : true
      const textMacth = expense.description
        .toLowerCase()
        .includes(text.toLowerCase())

      return startDateMatch && endDateMatch && textMacth
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createAt < b.createAt ? 1 : -1
      }

      if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1
      }
    })
}
