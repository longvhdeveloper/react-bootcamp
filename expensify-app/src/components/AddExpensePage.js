import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpeneseForm'
import { addExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
  addExpense = (expense) => {
    this.props.addExpense(expense)
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={(expense) => {
            this.addExpense(expense)
          }}
        />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpense(expense)),
  }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
