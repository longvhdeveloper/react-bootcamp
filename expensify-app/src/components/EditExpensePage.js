import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpeneseForm'
import { editExpense, removeExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
  editExpense = (updated) => {
    const id = this.props.expense.id
    this.props.editExpense(id, updated)
    this.props.history.push('/')
  }

  removeExpense = () => {
    const id = this.props.expense.id
    this.props.removeExpense({ id })
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.editExpense} />
        <button onClick={this.removeExpense}>Remove</button>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpense: (id, updated) => dispatch(editExpense(id, updated)),
    removeExpense: () => dispatch(removeExpense()),
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id
    }),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
