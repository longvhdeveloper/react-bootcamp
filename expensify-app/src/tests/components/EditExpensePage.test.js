import React from 'react'
import { shallow } from 'enzyme'
import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'
import { expect } from '@jest/globals'

let editExpense, removeExpense, history, wrapper

beforeEach(() => {
  editExpense = jest.fn()
  removeExpense = jest.fn()
  history = { push: jest.fn() }
  wrapper = shallow(
    <EditExpensePage
      expense={expenses[0]}
      editExpense={editExpense}
      history={history}
      removeExpense={removeExpense}
    />
  )
})

test('should render EditExpensePage correct', () => {
  expect(wrapper).toMatchSnapshot()
})

test('handle edit expense onSubmit', () => {
  const updated = {
    ...expenses[0],
    description: 'updated description',
  }
  wrapper.find('ExpenseForm').prop('onSubmit')(updated)
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, updated)
})

test('handle remove expense onClick', () => {
  wrapper.find('button').prop('onClick')()
  expect(history.push).toHaveBeenLastCalledWith('/')
  expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[0].id })
})
