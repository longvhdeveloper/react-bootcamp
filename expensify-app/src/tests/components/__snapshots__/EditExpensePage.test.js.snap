// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correct 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 195,
        "createAt": 0,
        "description": "Dum",
        "id": "1",
        "notes": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
